modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -480,6 +495,101 @@ inline IE::Blob::Ptr extractBlob(IECallContext& ctx, std::size_t i) {
 }
 } // anonymous namespace
 
+class cv::gimpl::ie::RequestPool {
+public:
+    using CompiledPtr = std::shared_ptr<cv::gimpl::ie::IECompiled>;
+    using Run      = std::function<void(cv::gimpl::ie::IECompiled&)>;
+    using Callback = std::function<void(cv::gimpl::ie::IECompiled&)>;
+    struct Task {
+        Run run;
+        Callback callback;
+    };
+
+    RequestPool(const IEUnit& uu)
+      : m_ndone(0u), m_nireq(uu.params.nireq) {
+          // FIXME: What's the capacity should be here (not bounded ??? )
+          m_tasks.set_capacity(m_nireq);
+          for (size_t i = 0; i < m_nireq; ++i) {
+              m_compileds.push_back(uu.compile());
+          }
+
+          m_worker = std::thread([this](){
+              this->worker();
+          });
+      }
+
+    void execute(Task&& t) {
+        m_tasks.push(Cmd{std::move(t)});
+    }
+
+    void wait_and_shutdown() {
+      for (size_t i = 0; i < m_nireq; ++i) {
+          m_tasks.push(Cmd{Stop{}});
+      }
+      m_worker.join();
+    }
+
+    ~RequestPool() {
+      // FIXME: What nobody calls wait_and_shutdown
+      //wait_and_shutdown();
+    }
+
+private:
+    struct Stop { };
+    using Cmd = cv::util::variant<Task, Stop>;
+
+    void Handler(Task task, CompiledPtr compiled) {
+        task.callback(*compiled);
+
+        Cmd cmd;
+        m_tasks.pop(cmd);
+
+        if (cv::util::holds_alternative<Stop>(cmd)) {
+            std::lock_guard<std::mutex> lock{m_mutex};
+            m_ndone++;
+            if (m_ndone == m_nireq) {
+                m_cv.notify_one();
+            }
+            return;
+        }
+
+        auto&& next_task = cv::util::get<Task>(cmd);
+        compiled->this_request.SetCompletionCallback(
+                std::bind(&cv::gimpl::ie::RequestPool::Handler, this, next_task, compiled));
+
+        next_task.run(*compiled);
+    }
+
+    void worker() {
+        for (auto&& compiled : m_compileds) {
+            Cmd cmd;
+            m_tasks.pop(cmd);
+
+            if (cv::util::holds_alternative<Stop>(cmd)) {
+                std::lock_guard<std::mutex> lock{m_mutex};
+                m_ndone++;
+                continue;
+            }
+
+            auto&& task = cv::util::get<Task>(cmd);
+            compiled->this_request.SetCompletionCallback(
+                    std::bind(&cv::gimpl::ie::RequestPool::Handler, this, task, compiled));
+
+            task.run(*compiled);
+        }
+        std::unique_lock<std::mutex> lock{m_mutex};
+        m_cv.wait(lock, [this]() { return m_ndone == m_nireq; });
+    }
+
+    size_t                   m_ndone;
+    const size_t             m_nireq;
+    std::vector<CompiledPtr> m_compileds;
+    QueueType<Cmd>           m_tasks;
+    std::thread              m_worker;
+    std::mutex               m_mutex;
+    std::condition_variable  m_cv;
+};
```

Comment:
Please separate class declaration from its implementation. Move methods out. This class is hard to review.

Reply:
Will not dive deeper before that.

Reply:
Done